### ISSUE TITLE
Apollo Kotlin 3 subscription collect callback is never called

### ISSUE THREAD

**Issue by simondorociak** [apollographql/apollo-kotlin#3791]
**Summary**
Launched subscription not collects new values (I do not receive callbacks in Activity)

**Version**
3.0.0-beta03

We have two subscriptions in app. One is chat (in Fragment) that handles receiving messages and second one (new) we try to implement handles overall app "unread" states of sections. Subscription in chat worked. After I integrated second subscription that second does not work. Im launching that new apollo subscription like this (its called in MainActivity.kt in our single-activity app because it has to be collected during app lifecycle):

```
lifecycleScope.launch {
            client.subscribe(UpdateDataCountSubscription()).execute()
                .flowWithLifecycle(lifecycle, Lifecycle.State.STARTED)
                .retryWhen { cause, _ ->
                    Timber.d("Reconnecting...")
                    Timber.e(cause)
                    delay(10000L)
                    true
                }
                .collect {
                    Timber.d("Subscribed to app data states")
                }
        }
```

But collect is never called. I see in logs that code in **SocketSubscriptionTransport** is called correctly:

```
override fun operationResponse(id: String, payload: Map<String, Any?>) {
        val result = mutableEvents.tryEmit(OperationResponse(id, payload))
        Timber.d("operationResponse -> $result")
    }
```

this returns true and then result is propagated into 

```
override fun <D : Operation.Data> execute(
        request: ApolloRequest<D>,
    ): Flow<ApolloResponse<D>>

```
And here is correctly handled by map function:

```
.map {
            when (it) {
                is OperationResponse -> {
                    Timber.d("OperationResponse parsing")
                    ResponseBodyParser.parse(
                        it.payload,
                        request.operation,
                        request.executionContext[CustomScalarAdapters]!!
                    )
                }
                is OperationError -> throw ApolloNetworkException("Cannot start operation ${request.operation.name()}: ${it.payload}")
                is NetworkError -> throw ApolloNetworkException("Network error while executing ${request.operation.name()}", it.cause)

                // Cannot happen as these events are filtered out upstream
                is OperationComplete, is GeneralError -> error("Unexpected event $it")
            }
        }
```

So I see in logs **OperationResponse parsing** message but this is last thing I see in log. After that I should get callback in collect but Im not receiving it, any error so I don't know where value is. When i try to invoke from server second time subscription Im receiving obviously error that Flow contains more than one element.

Version Im using is: **3.0.0-beta03** This is feature blocker for our app so any help would be useful how to make that work. Thanks.

**Comment by martinbonnin** [comment]
Thanks for reaching out! Can you try with `3.0.0`? It has some API changes and it'll be easier to investigate on the latest version.

**Comment by simondorociak** [comment]
Seems after I updated to 3.0.0 I cant anymore override SocketSubscriptionTransport class. I have to override this because I have to pass JWT token when Im connecting to socket. Any suggestion please?

**Comment by martinbonnin** [comment]
If you want to pass a payload to the 'connection_init' message, you can use `connectionPayload`:

```kotlin
val apolloClient = ApolloClient.Builder()
    .httpServerUrl("http://localhost:8080/graphql")
    .webSocketServerUrl("http://localhost:8080/subscriptions")
    .wsProtocol(
        SubscriptionWsProtocol.Factory(
            connectionPayload = {
              mapOf("token" to upToDateToken)
            }
        )
    )
    .build()
```

If you need more control over the WebSocket handshake itself, your best bet is to implement your own `WsProtocol`. I'm on mobile right now but can follow later with more details

**Comment by simondorociak** [comment]
Thank you. I will try that. I now changed version to 3.0.0 but this error gets shown even I cleared project also invalidated cash of IDE:
```

Execution failed for task ':app:checkApolloVersions'.
> ApolloGraphQL: All apollo versions should be the same. Found:
  [3.0.0, 3.0.0-beta03]
```

This is reallly od. I dont have any beta03 number in the project.

**Comment by martinbonnin** [comment]
Maybe from the plugin or something like this?

**Comment by simondorociak** [comment]
@martinbonnin Yes it was caused by plugin. You did a little bit bigger changes from beta do production/staging so I have to rewrite besically whole repository layer which I dont like. I did not except such a huge changes (imho Optional wrapper of nullable fields is redundantly complex I really liked kotlin nullable params). I will get back you when I will rewrite repositories.

**Comment by martinbonnin** [comment]
> Optional wrapper of nullable fields is redundantly complex I really liked kotlin nullable params

That's a long story and the fact that GraphQL does not distinguish between nullable and optional doesn't help but for variables, you might be interested in `generateOptionalOperationVariables`:

```kotlin
apollo {
  generateOptionalOperationVariables.set(false)
}
```

This will map all optional variables to Kotlin nullable parameters. Maybe that can make your callsites a bit cleaner.

**Comment by simondorociak** [comment]
@martinbonnin  I migrated to 3.0.0 and error occurs as in my first post :/ My subscription:

```
lifecycleScope.launch {
            client.subscription(UpdateDataCountSubscription()).toFlow()
                .flowWithLifecycle(lifecycle, Lifecycle.State.STARTED)
                .retryWhen { cause, _ ->
                    Timber.d("Reconnecting...")
                    Timber.e(cause)
                    delay(10000L)
                    true
                }
                .collect {
                    Timber.d("Subscribed to app data states")
                }
        }
```

Doesnt receive collect callback. And when I invoked from backend second time I got error that Flow has more than one element.

This is configuration of Apollo:

@Provides
    @Singleton
    fun provideApolloClient(
            application: Application,
            prefs: SharedPrefsManager,
            httpClient: OkHttpClient) : ApolloClient {
        return ApolloClient.Builder()
            .networkTransport(
                HttpNetworkTransport.Builder()
                    .serverUrl(BuildConfig.SERVER_URL)
                    .okHttpClient(httpClient)
                    .build()

            )
            .webSocketServerUrl(BuildConfig.SOCKET_URL)
            .wsProtocol(AppSubscriptionWsProtocol.Factory(
                connectionPayload = {
                    mapOf("Authorization" to "JWT ${prefs.getJwtToken()}")
                }
            ))
            .addInterceptor(RefreshTokenInterceptor(application, prefs))
            .build()
    }

One subscription placed in chat fragment works but this subscription placed in MainActivity doesnt work.

**Comment by martinbonnin** [comment]
Thanks for providing that! 2 ways to move forward:

1. Is your backend publicly available by any chance? I could try to see if I can reproduce the error.
2. Can you run your app with a proxy like [Charles](https://www.charlesproxy.com/) to dump the network traffic and see if anything goes wrong?

**Comment by simondorociak** [comment]
@martinbonnin Sorry for taking your time. Im an idiot. 🤦 I now after I copy pasted code I now checked my AppInterceptor and here was code:

override fun <D : Operation.Data> intercept(
        request: ApolloRequest<D>,
        chain: ApolloInterceptorChain
    ): Flow<ApolloResponse<D>> {
        val operationName = request.operation.name()
        Timber.d("Operation name: $operationName")
        if (operationName.contains(OPERATION_SUBSCRIPTION, true)) {
            Timber.d("Subscription request detected -> do not intercept")
            return chain.proceed(request)
        } else {
            return flow {
               // handling 401 code
           }

Here was problem that in case of subscription I have to not process request otherwise subscription is broken. Here Im checking operation name and if its subscription Im ignoring it. But I named second subscription incorrectly and that caused problem because I processed subscription (it was handled by apollo interceptor and flow was thrown away and was executed again and actually callback was lost). 

Sorry again. Now it works like a charm. At least i migrated to stable version and also code of your socket looks much more cleaner thank you.
